extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInvalidateInterceptor.java

```diff
@@ -25,20 +31,58 @@ public Object intercept(InvocationContext invocationContext) throws Exception {
         if (interceptionContext.getInterceptorBindings().isEmpty()) {
             // This should never happen.
             LOGGER.warn(INTERCEPTOR_BINDINGS_ERROR_MSG);
+            return invocationContext.proceed();
+        } else if (isUniReturnType(invocationContext)) {
+            return invalidateNonBlocking(invocationContext, interceptionContext);
         } else {
-            Object key = null;
-            for (CacheInvalidate binding : interceptionContext.getInterceptorBindings()) {
-                Cache cache = cacheManager.getCache(binding.cacheName()).get();
-                if (key == null) {
-                    key = getCacheKey(cache, interceptionContext.getCacheKeyParameterPositions(),
-                            invocationContext.getParameters());
-                }
-                if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debugf("Invalidating entry with key [%s] from cache [%s]", key, binding.cacheName());
-                }
-                cache.invalidate(key).await().indefinitely();
-            }
+            return invalidateBlocking(invocationContext, interceptionContext);
+        }
+    }
+
+    private Object invalidateNonBlocking(InvocationContext invocationContext,
+            CacheInterceptionContext<CacheInvalidate> interceptionContext) {
+        // Because of this array, the key will only be built once and reused by each iteration of the following Multi.
+        Object[] key = new Object[1];
```

Comment:
Could you be more specific? Why the array? Is it just a holder of a reference that can be only set on the same thread?

Reply:
Yes it's a holder reference to "hack" the lambda restriction about final variables. The idea is to avoid recalculating the same key each time an invalidation is performed because the `key` value will always be the same.

I'm not entirely sure, but the interceptor execution may involve several threads as `Merge` implies concurrency in Mutiny. But the key value will always be the same and if it's not available to a thread (`null`) then it will be calculated.

Reply:
After re-reading the extension code, I think there's an issue (unrelated with the holder reference) with reusing the same key...

Reply:
I just fixed a bug with the reusable cache key during an invalidation.

When a method annotated with `@CacheInvalidate` does not have any arguments, the cache name is used to [build a default key](https://github.com/quarkusio/quarkus/blob/e0bb1503fbf0091538470424270a5853b8a2785f/extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java#L110-L112). Reusing the same key means that the following code won't work as one of the caches won't be invalidated. It's not a common scenario as most cached methods have arguments, but still a bug 😃 

```
@CacheInvalidate(cacheName = "foo")
@CacheInvalidate(cacheName = "bar")
public void invalidate() {
}
```

Thanks for questioning the holder reference, it made me realize this 😄 The hack is gone now.

Reply:
I'll add a test for this case in a few minutes.

Reply:
Test added.

Reply:
Great!